# Data Analysis for Exoplanets

This notebook is ment to perform some data analysis on [exoplanets](https://en.wikipedia.org/wiki/Exoplanet). The dataset is from The [NASA Exoplanet Archive](https://exoplanetarchive.ipac.caltech.edu/cgi-bin/TblView/nph-tblView?app=ExoTbls&config=planets).

### Imports

In [24]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [25]:
'''
Function to load in the data and print the header which explains what each of the column fields mean
'''
def load_planets_csv(csvFilename, headerlines=76):
    with open(csvFilename, 'r') as csvFile:
        linecounter = 1
        for line in csvFile:
            print(line)
            linecounter+=1
            if linecounter > headerlines:
                break
    return pd.read_csv(csvFilename, header=headerlines)

In [26]:
# Load in the datafile and inspect the column descriptions:
planet_df = load_planets_csv("./planets_2020.07.25_10.13.04.csv")

# This file was produced by the NASA Exoplanet Archive  http://exoplanetarchive.ipac.caltech.edu

# Sat Jul 25 10:13:04 2020

#

# COLUMN pl_hostname:    Host Name

# COLUMN pl_letter:      Planet Letter

# COLUMN pl_name:        Planet Name

# COLUMN pl_discmethod:  Discovery Method

# COLUMN pl_controvflag: Controversial Flag

# COLUMN pl_pnum:        Number of Planets in System

# COLUMN pl_orbper:      Orbital Period [days]

# COLUMN pl_orbpererr1:  Orbital Period Upper Unc. [days]

# COLUMN pl_orbpererr2:  Orbital Period Lower Unc. [days]

# COLUMN pl_orbperlim:   Orbital Period Limit Flag

# COLUMN pl_orbsmax:     Orbit Semi-Major Axis [au])

# COLUMN pl_orbsmaxerr1: Orbit Semi-Major Axis Upper Unc. [au]

# COLUMN pl_orbsmaxerr2: Orbit Semi-Major Axis Lower Unc. [au]

# COLUMN pl_orbsmaxlim:  Orbit Semi-Major Axis Limit Flag

# COLUMN pl_orbeccen:    Eccentricity

# COLUMN pl_orbeccenerr1: Eccentricity Upper Unc.

# COLUMN pl_orbeccenerr2: Eccentricity Lower Unc.

# COLUMN pl_orb

In [27]:
# Look at the resulting pandas dataframe
planet_df

pl_hostname pl_letter     pl_name    pl_discmethod  pl_controvflag  \
0         11 Com         b    11 Com b  Radial Velocity               0   
1         11 UMi         b    11 UMi b  Radial Velocity               0   
2         14 And         b    14 And b  Radial Velocity               0   
3         14 Her         b    14 Her b  Radial Velocity               0   
4       16 Cyg B         b  16 Cyg B b  Radial Velocity               0   
...          ...       ...         ...              ...             ...   
4192     tau Gem         b   tau Gem b  Radial Velocity               0   
4193     ups And         b   ups And b  Radial Velocity               0   
4194     ups And         c   ups And c  Radial Velocity               0   
4195     ups And         d   ups And d  Radial Velocity               0   
4196      xi Aql         b    xi Aql b  Radial Velocity               0   

      pl_pnum    pl_orbper  pl_orbpererr1  pl_orbpererr2  pl_orbperlim  ...  \
0           1   326.030000       0.320000      -0.320000           0.0  ...   
1           1   516.219970       3.200000      -3.200000           0.0  ...   
2           1   185.840000       0.230000      -0.230000           0.0  ...   
3           1  1773.400020       2.500000      -2.500000           0.0  ...   
4           1   798.500000       1.000000      -1.000000           0.0  ...   
...       ...          ...            ...            ...           ...  ...   
4192        1   305.500000       0.100000      -0.100000           0.0  ...   
4193        3     4.617033       0.000023      -0.000023           0.0  ...   
4194        3   241.258000       0.064000      -0.064000           0.0  ...   
4195        3  1276.460000       0.570000      -0.570000           0.0  ...   
4196        1   136.750000       0.250000      -0.250000           0.0  ...   

      st_mass  st_masserr1  st_masserr2  st_masslim  st_rad  st_raderr1  \
0        2.70         0.30        -0.30         0.0   19.00        2.00   
1        2.78         0.69        -0.69         0.0   29.79        2.84   
2        2.20         0.10        -0.20         0.0   11.00        1.00   
3        0.90         0.04        -0.04         0.0    0.93        0.01   
4        1.08         0.04        -0.04         0.0    1.13        0.01   
...       ...          ...          ...         ...     ...         ...   
4192     2.30         0.30        -0.30         0.0   26.80        0.70   
4193     1.30          NaN          NaN         0.0    1.56         NaN   
4194     1.30          NaN          NaN         0.0    1.56         NaN   
4195     1.30          NaN          NaN         0.0    1.56         NaN   
4196     2.20          NaN          NaN         0.0   12.00         NaN   

      st_raderr2  st_radlim   rowupdate  \
0          -2.00        0.0  2014-05-14   
1          -2.84        0.0  2018-09-06   
2          -1.00        0.0  2014-05-14   
3          -0.01        0.0  2018-09-06   
4          -0.01        0.0  2018-09-06   
...          ...        ...         ...   
4192       -0.70        0.0  2014-05-14   
4193         NaN        0.0  2019-01-31   
4194         NaN        0.0  2019-01-31   
4195         NaN        0.0  2019-01-31   
4196         NaN        0.0  2014-05-14   

                                 pl_facility  
0                           Xinglong Station  
1     Thueringer Landessternwarte Tautenburg  
2          Okayama Astrophysical Observatory  
3                     W. M. Keck Observatory  
4                     Multiple Observatories  
...                                      ...  
4192                        Lick Observatory  
4193                        Lick Observatory  
4194                  Multiple Observatories  
4195                  Multiple Observatories  
4196       Okayama Astrophysical Observatory  

[4197 rows x 72 columns]

From the above cell, we can see that there are 4,197 confrimed planets in this dataset, with each row in the data representing a planet. Let's group the planets by their host star names and see how many confirmed planets each star has.

In [17]:
planet_df.groupby('pl_hostname').size().describe()

count    3115.000000
mean        1.347352
std         0.764953
min         1.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         8.000000
dtype: float64

There are 3,115 distinct [planetary systems](https://en.wikipedia.org/wiki/Planetary_system). We notice that most of the planetary systems consist only of a single planet. The average system has 1.3 planets and the system with the most planets has 8! This system known as KOI-351 or [Kepler-90](https://en.wikipedia.org/wiki/Kepler-90) had its [eigth planet](https://en.wikipedia.org/wiki/Kepler-90i) discovered in 2017 [using machine learning methods](https://arxiv.org/abs/1712.05044) developed at Google. Let's inspect that system more closely.

In [29]:
# Sort the systems by number of planets in decreasing order:
planet_df.groupby('pl_hostname').size().sort_values(ascending=False)

pl_hostname
KOI-351       8
TRAPPIST-1    7
Kepler-20     6
Kepler-80     6
HD 10180      6
             ..
Kepler-428    1
Kepler-427    1
Kepler-426    1
Kepler-425    1
11 Com        1
Length: 3115, dtype: int64

In [30]:
# Select the part of the dataframe corresponding to the host KOI-351. This gives us the eigth planets around that star
planet_df[planet_df['pl_hostname'] == 'KOI-351']

pl_hostname pl_letter      pl_name pl_discmethod  pl_controvflag  \
1405     KOI-351         b    KOI-351 b       Transit               0   
1406     KOI-351         c    KOI-351 c       Transit               0   
1407     KOI-351         d    KOI-351 d       Transit               0   
1408     KOI-351         e    KOI-351 e       Transit               0   
1409     KOI-351         f    KOI-351 f       Transit               0   
1410     KOI-351         g    KOI-351 g       Transit               0   
1411     KOI-351         h    KOI-351 h       Transit               0   
3617     KOI-351         i  Kepler-90 i       Transit               0   

      pl_pnum   pl_orbper  pl_orbpererr1  pl_orbpererr2  pl_orbperlim  ...  \
1405        8    7.008151       0.000019      -0.000019           0.0  ...   
1406        8    8.719375       0.000027      -0.000027           0.0  ...   
1407        8   59.736670       0.000380      -0.000380           0.0  ...   
1408        8   91.939130       0.000730      -0.000730           0.0  ...   
1409        8  124.914400       0.001900      -0.001900           0.0  ...   
1410        8  210.606970       0.000430      -0.000430           NaN  ...   
1411        8  331.600590       0.000370      -0.000370           0.0  ...   
3617        8   14.449120       0.000200      -0.000200           0.0  ...   

      st_mass  st_masserr1  st_masserr2  st_masslim  st_rad  st_raderr1  \
1405      1.2          0.1         -0.1         0.0     1.2         0.1   
1406      1.2          0.1         -0.1         0.0     1.2         0.1   
1407      1.2          0.1         -0.1         0.0     1.2         0.1   
1408      1.2          0.1         -0.1         0.0     1.2         0.1   
1409      1.2          0.1         -0.1         0.0     1.2         0.1   
1410      1.2          0.1         -0.1         0.0     1.2         0.1   
1411      1.2          0.1         -0.1         0.0     1.2         0.1   
3617      1.2          0.1         -0.1         0.0     1.2         0.1   

      st_raderr2  st_radlim   rowupdate  pl_facility  
1405        -0.1        0.0  2014-05-14       Kepler  
1406        -0.1        0.0  2014-05-14       Kepler  
1407        -0.1        0.0  2014-05-14       Kepler  
1408        -0.1        0.0  2014-05-14       Kepler  
1409        -0.1        0.0  2014-05-14       Kepler  
1410        -0.1        0.0  2014-05-14       Kepler  
1411        -0.1        0.0  2014-05-14       Kepler  
3617        -0.1        0.0  2017-12-14       Kepler  

[8 rows x 72 columns]

In [32]:
# 
planet_df[planet_df['pl_hostname'] == 'KOI-351']['pl_orbsmax']

1405    0.074
1406    0.089
1407    0.320
1408    0.420
1409    0.480
1410    0.710
1411    1.010
3617      NaN
Name: pl_orbsmax, dtype: float64